### PR TITLE
Add tests for sub_group_mask non const member functions

### PR DESCRIPTION
**Pull request by vasilytric**
`sub_group_mask` it's an extension, documentations can be found here: [intel/llvm/doc/SubGroupMask](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SubGroupMask/SubGroupMask.asciidoc)

This PR depends on https://github.com/KhronosGroup/SYCL-CTS/pull/200
`sub_group_mask_common.h` file can be found in PR#200

### CODE REVIEW

**Comment by keryell** [comment]
Let's wait for https://github.com/KhronosGroup/SYCL-CTS/pull/200 to settle first.

**Review by keryell — CHANGES_REQUESTED** [review]

File: tests/sub_group_mask/sub_group_mask_flip_api.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+struct check_result_flip {
+  bool operator()(sycl::ext::oneapi::sub_group_mask &sub_group_mask,
+                  const sycl::sub_group &sub_group) {
+    unsigned long before_flip, after_flip;
+    unsigned long zero = 0;
+    sub_group_mask.extract_bits(before_flip);
+    sub_group_mask.flip();
+    sub_group_mask.extract_bits(after_flip);
+    return ((before_flip ^ after_flip) == ~zero);
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/sub_group_mask/sub_group_mask_flip_api.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/204/commits/e0e16fd8448171d1f08e22319ccdfb60e1a4bddd#diff-d1406ca25ff16af0b829e9a90174b01dcdd9f99ab13e22500ecbdc808388982bR25)

File: tests/sub_group_mask/sub_group_mask_flip_api.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+namespace TEST_NAMESPACE {
+
+using namespace sycl_cts;
+#ifdef SYCL_EXT_ONEAPI_SUB_GROUP_MASK
+
+struct check_result_flip {
+  bool operator()(sycl::ext::oneapi::sub_group_mask &sub_group_mask,
+                  const sycl::sub_group &sub_group) {
+    unsigned long before_flip, after_flip;
+    unsigned long zero = 0;
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/sub_group_mask/sub_group_mask_flip_api.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/204/commits/e0e16fd8448171d1f08e22319ccdfb60e1a4bddd#diff-d1406ca25ff16af0b829e9a90174b01dcdd9f99ab13e22500ecbdc808388982bL22)

File: tests/sub_group_mask/sub_group_mask_flip_id.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+struct check_result_flip_id {
+  bool operator()(sycl::ext::oneapi::sub_group_mask &sub_group_mask,
+                  const sycl::sub_group &sub_group) {
+    auto local_id = sub_group.get_local_id();
+    unsigned long before_flip, after_flip;
+    sub_group_mask.extract_bits(before_flip);
+    sub_group_mask.flip(local_id);
+    sub_group_mask.extract_bits(after_flip);
+    return ((before_flip ^ after_flip) == 1 << local_id.get(0));
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/sub_group_mask/sub_group_mask_flip_id.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/204/commits/e0e16fd8448171d1f08e22319ccdfb60e1a4bddd#diff-cd29a5aef461102303f5e5a6bbbc310d462415fd23fbbb2097d3afd60ab2d68aL26-L27)

File: tests/sub_group_mask/sub_group_mask_set_api.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+#ifdef SYCL_EXT_ONEAPI_SUB_GROUP_MASK
+
+struct check_result_set {
+  bool operator()(sycl::ext::oneapi::sub_group_mask &sub_group_mask,
+                  const sycl::sub_group &sub_group) {
+    unsigned long after_set;
+    unsigned long zero = 0;
+    sub_group_mask.set();
+    sub_group_mask.extract_bits(after_set);
+    return (after_set == ~zero);
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/sub_group_mask/sub_group_mask_set_api.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/204/commits/e0e16fd8448171d1f08e22319ccdfb60e1a4bddd#diff-fd207f571e1fb3ff3a62fa729e3085e112aec215995fd5ca82575f5bfd9d6fb5L25-L26)

File: tests/sub_group_mask/sub_group_mask_reset_low.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+struct check_result_reset_low {
+  bool operator()(sycl::ext::oneapi::sub_group_mask &sub_group_mask,
+                  const sycl::sub_group &sub_group) {
+    auto low = sub_group_mask.find_low();
+    unsigned long after_reset, before_reset;
+    sub_group_mask.extract_bits(before_reset);
+    sub_group_mask.reset_low();
+    sub_group_mask.extract_bits(after_reset);
+    return ((after_reset ^ before_reset) == 1 << low);
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/sub_group_mask/sub_group_mask_reset_low.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/204/commits/e0e16fd8448171d1f08e22319ccdfb60e1a4bddd#diff-7538ac13065e2c4e4b0d3e836b62afdbce0dd7335a1b33034a60decd70345077L26-L27)

File: tests/sub_group_mask/sub_group_mask_reset_id.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                  const sycl::sub_group &sub_group) {
+    for (int N = 0; N < sub_group_mask.size(); N++) {
+      if (N % 3 == 0) sub_group_mask.reset(sycl::id(N));
+    }
+
+    for (int N = 0; N < sub_group_mask.size(); N++) {
+      switch (N % 3) {
+        case 0:
+          if (sub_group_mask.test(sycl::id(N))) return false;
+          break;
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/sub_group_mask/sub_group_mask_reset_id.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/204/commits/e0e16fd8448171d1f08e22319ccdfb60e1a4bddd#diff-d8d6ab904bdaf32fee120080f848f249ca7fd8fb7955c9022d4efe09fc79d939L29-L30)

File: tests/sub_group_mask/sub_group_mask_reset_api.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+using namespace sycl_cts;
+#ifdef SYCL_EXT_ONEAPI_SUB_GROUP_MASK
+
+struct check_result_reset {
+  bool operator()(sycl::ext::oneapi::sub_group_mask &sub_group_mask,
+                  const sycl::sub_group &sub_group) {
+    unsigned long after_reset;
+    sub_group_mask.reset();
+    sub_group_mask.extract_bits(after_reset);
+    return (after_reset == 0);
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/sub_group_mask/sub_group_mask_reset_api.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/204/commits/e0e16fd8448171d1f08e22319ccdfb60e1a4bddd#diff-2fc771cf626670244d53a9352068abcb1a48fd10ba64e7e30b5ab835c59f1a44L24-L25)

File: tests/sub_group_mask/sub_group_mask_reset_id.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+using namespace sycl_cts;
+#ifdef SYCL_EXT_ONEAPI_SUB_GROUP_MASK
+
+struct check_result_reset_id {
+  bool operator()(sycl::ext::oneapi::sub_group_mask &sub_group_mask,
+                  const sycl::sub_group &sub_group) {
+    for (int N = 0; N < sub_group_mask.size(); N++) {
+      if (N % 3 == 0) sub_group_mask.reset(sycl::id(N));
+    }
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/sub_group_mask/sub_group_mask_reset_id.cpp]: > I am trying to understand your code obfuscation contest :-)

Oh, unfortunately, some statement of these test could be so obfuscated

I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/204/commits/e0e16fd8448171d1f08e22319ccdfb60e1a4bddd#diff-d8d6ab904bdaf32fee120080f848f249ca7fd8fb7955c9022d4efe09fc79d939L21-L23)

**Comment by vasilytric** [comment]
@keryell  I am sorry for the force-push
I decide to do it due to the fact my previous push contained unintended commits

File: tests/sub_group_mask/sub_group_mask_reset_high.cpp [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+struct check_result_reset_high {
+  bool operator()(sycl::ext::oneapi::sub_group_mask &sub_group_mask,
+                  const sycl::sub_group &sub_group) {
+    auto high = sub_group_mask.find_high();
+    unsigned long after_reset, before_reset;
+    sub_group_mask.extract_bits(before_reset);
+    sub_group_mask.reset_high();
+    sub_group_mask.extract_bits(after_reset);
+    return ((after_reset ^ before_reset) == 1 << high);
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/sub_group_mask/sub_group_mask_reset_high.cpp]: Comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/204/commits/e0e16fd8448171d1f08e22319ccdfb60e1a4bddd#diff-001baad40d55a09bfbd12a47c7400f5fe59dee60da9ec86e2ec75ba46f5f889bR26)

**Review by keryell — COMMENTED** [review]
It should be good when it is merged from main branch to move into subdirectories, once https://github.com/KhronosGroup/SYCL-CTS/pull/200 has landed.

**Review by keryell — APPROVED** [review]
Good!

**Pull request merged**